book/src/leader-validator-transition.md

```diff
@@ -0,0 +1,157 @@
+
+## Leader
+
+A leader operates on a single fork for a specific PoH slot.
+
+
+## PoH
+
+Two different objects for managing PoH for leaders and validators.  Both only
+work for a specific range of ticks and will error out and send an exit signal
+when they are done.
+
+### PoH Generator
+
+This object handles PoH generation for a Validator.  It solves the following
+problems for validators
+
+1. Keep track of *time* as defined by PoH height, and stop the validator when it
+hits the scheduled leader slot.
+
+2. Reset the clock whenever a validator votes.  The last validator vote is the
+validators most recent commitment. It is therefore the closest fork to the
+scheduled slot.
+
+3. Provide the entries that connect the last vote to the scheduled slot.
+
+``` PohGenerator {
+    /// Start the PoH thread, and compute it until end_ticks.
+    /// Once computed, the exit_signal is sent
+    pub fn new(
+        start: &Hash,
+        start_ticks: u64,
+        end_ticks: u64,
+        exit_signal: Sender<()>);
+
+    /// The is called when a validator votes and needs to reset PoH with its
+    /// vote.  Error is returned if end_ticks is reached first.
+    pub fn reset(last_id: &Hash, reset_height: u64) -> Result<()>;
+
+    /// The list of entries from `start` or `reset_height` to `end_ticks`.
+    /// Error is returned if end_ticks has not been reached yet.
+    pub fn final_entries() -> Result<Vec<Entry>>;
+} ```
+
+### PoH Recorder
+
+This object handles PoH recording for a Leader.  It solves the following
+problems:
+
+1. Keep track of *time* as defined by PoH height, and stop the leader when it
+gets to the end of its scheduled leader slot.
+
+2. Register ticks with the BankState as they are produced by PoH.
+
+3. Record entries that are produced by the TPU into PoH.
+
+```
+PohRecorder {
+    /// Start the PoH thread, and compute it until end_ticks.
+    /// The recorder will continue to call register tick on the BankState.
+    /// TODO: we can remove `register_ticks` on the bank if the slot ticks are
+    /// the only valid ticks.
+    /// Once computed, the exit_signal is sent.
+    pub fn new(
+	    entries: &[Entry],
+        end_ticks: u64,
+        sender: Sender<Vec<Entry>>,
+	    bank_state: Arc<BankState>,
+        exit_signal: Sender<()>);
+
+    /// Record transactions /// Error is returned if end_ticks is reached first
+    pub fn record(&self, mixin: Hash, txs: Vec<Transaction>) -> Result<()>;
+} 
+```
+
+
+## Fullnode Loop
+
+This object manages the transition between modes.  The main idea is that once a
+ledger is replayed, the validator can run until there is proof that it can be a
+leader.  The Leader can then run and record its transactions.
+
+``` 
+/// * forks - the Forks generated by replaying the ledger
+/// * leader_scheduler - the leader_scheduler generated by replaying the ledger 
+/// * ncp -  the network
+
+let main(forks: Forks, leader_scheduler: LeaderScheduler, ncp: Ncp) {
```

Comment:
Bizarre to see this Rust-ish code. It clearly wouldn't compile and it references terms that don't exist in the codebase, such as `Ncp` and `Validator`. Why are you asking the reader to make sense of this?

Reply:
@garious it's just some psudocode to give you structure of the loop.  the main thing is that the loop is synchronized on PoH and does a full start and stop of validator and leader